include/Mat.h

```diff
@@ -1169,6 +1338,250 @@ class Const_MatIter{
     }
 };
 
+template<class Type, class Type2>
+class iMat{
+    public:
+    Mat<Type> matrix;
+    Mat<Type2> index;
+    
+    iMat(Mat<Type>& m, const Mat<Type2>& i) : matrix(m), index(i){
+        if(std::is_same<Type2, bool>::value){
+            for(long i = 0; i < matrix.ndim; i++){
+                if(index.dims[i] != matrix.dims[i])
+                    throw invalid_argument("mask index broadcasting not yet implemented");
+            }
+        }
+        else if(index.ndim != 1)
+            throw invalid_argument("index lists with ndim != 1 not yet implemented");
+    }
+
+    Mat<Type>& operator=(const Mat<Type>& b){
+        /*if(b.dims[0] != index.dims[0]) throw invalid_argument
+            ("broadcasting assignment to an indexed matrix not yet implemented");*/
+        if(matrix.memory == b.memory){
+            return *this = b.copy();
+        }
+        MatIter<Type> j = matrix.begin();
+        Const_MatIter<Type> k = b.begin();
+        if(std::is_same<Type2, bool>::value){
+            if(index.ndim > 1 && b.ndim > 1) throw invalid_argument
+                    ("assignment to indexed array requires one operand to be 1d");
+            for(auto i : index){
+                if(i){
+                    *j = *k;
+                    ++k;
+                }
```

Comment:
Can check if k reaches the end by the end of the loop for boolean indexing. Not sure if we should check bounds as we go for j